File: src/packages/protocol/src/events.ts
// createEvent / signEvent / verifyEvent.
// Per docs/spec/03-modules-platform.md §1 and ADR-003.
// Ported from legacy/src-v1/packages/protocol/src/events.ts:29-134.
import { ed25519 } from '@noble/curves/ed25519'
import type { AgentKey, IntentPayload, SerendipEvent, SerendipKind } from './types.js'
import { canonicalize, sha256hex } from './canonical.js'
import { InvalidKindError, PayloadTooLargeError } from './errors.js'
import { bytesToHex, hexToBytes } from './utils.js'

/**
 * 64 KiB payload limit. SPEC 02-data-model.md §1.1 and
 * 03-modules-platform.md §1 acceptance 2.
 */
export const MAX_PAYLOAD_BYTES = 65_536

const PROTOCOL_KINDS: ReadonlySet<SerendipKind> = new Set<SerendipKind>([
  'intent.broadcast',
  'intent.match',
  'intent.verify',
  'intent.subscribe',
  'identity.register',
  'identity.delegate',
  'identity.revoke',
])

function assertKind(kind: string): asserts kind is SerendipKind {
  if (!PROTOCOL_KINDS.has(kind as SerendipKind)) {
    throw new InvalidKindError(kind)
  }
}

/**
 * Construct an unsigned event envelope. The returned shape omits
 * `id`, `sig`, `pubkey`, and `operator_pubkey`; those are filled in
 * by signEvent from the signing AgentKey.
 */
export function createEvent(
  kind: SerendipKind,
  payload: IntentPayload,
  tags: string[],
  createdAt?: number,
): Omit<SerendipEvent, 'sig' | 'id' | 'pubkey' | 'operator_pubkey'> {
  assertKind(kind)
  return {
    v: 1,
    created_at: createdAt ?? Math.floor(Date.now() / 1000),
    kind,
    payload,
    tags,
    visibility: 'public',
  }
}

/**
 * Sign an unsigned event with an AgentKey. Throws:
 * - PayloadTooLargeError if the payload's UTF-8 JSON byte length
 *   exceeds MAX_PAYLOAD_BYTES (SPEC 02-data-model §1.1).
 * - InvalidKindError if the event's kind is not a protocol-layer kind
 *   (defence in depth; createEvent already checks).
 */
export async function signEvent(
  event: Omit<SerendipEvent, 'sig' | 'id' | 'pubkey' | 'operator_pubkey'>,
  agentKey: AgentKey,
): Promise<SerendipEvent> {
  assertKind(event.kind)

  const payloadBytes = new TextEncoder().encode(JSON.stringify(event.payload)).length
  if (payloadBytes > MAX_PAYLOAD_BYTES) {
    throw new PayloadTooLargeError(payloadBytes, MAX_PAYLOAD_BYTES)
  }

  const withKeys = {
    ...event,
    pubkey: agentKey.publicKey,
    operator_pubkey: agentKey.delegatedBy,
  }

  const canonical = canonicalize(withKeys)
  const id = sha256hex(canonical)
  const sig = bytesToHex(ed25519.sign(hexToBytes(id), agentKey.privateKey))

  return { ...withKeys, id, sig }
}

/**
 * Verify an event's id and signature. Never throws; returns false on
 * any mismatch or structural error. Per SPEC 03-modules-platform §1
 * acceptance 3, key expiry / revocation is NOT checked here — that is
 * the relay's responsibility.
 */
export async function verifyEvent(event: SerendipEvent): Promise<boolean> {
  try {
    const { id, sig, ...rest } = event

    const expectedId = sha256hex(canonicalize(rest))
    if (id !== expectedId) return false

    const idBytes = hexToBytes(id)
    const sigBytes = hexToBytes(sig)
    const pubkeyBytes = hexToBytes(event.pubkey)

    return ed25519.verify(sigBytes, idBytes, pubkeyBytes)
  } catch {
    return false
  }
}

File: packages/skill-hermes/agentxp_skill_hermes/events.py
"""create_event / sign_event / verify_event.

Matches the TypeScript protocol (`packages/protocol/src/events.ts`)
behavior under ADR-003. The only observable difference is that the
Python signer additionally cross-checks the canonical byte count
against a recomputation before signing — this is the typed error
demanded by SPEC 03-modules-product §4 acceptance case 3.
"""

from __future__ import annotations

import json
import time
from typing import Any, Optional

from .canonical import canonicalize, sha256_hex
from .keys import AgentKey, sign_bytes, verify_bytes

MAX_PAYLOAD_BYTES = 65_536

_PROTOCOL_KINDS = frozenset(
    {
        "intent.broadcast",
        "intent.match",
        "intent.verify",
        "intent.subscribe",
        "identity.register",
        "identity.delegate",
        "identity.revoke",
    }
)


class InvalidKindError(ValueError):
    def __init__(self, kind: str):
        super().__init__(f"invalid protocol kind: {kind}")
        self.kind = kind


class PayloadTooLargeError(ValueError):
    def __init__(self, actual: int, limit: int = MAX_PAYLOAD_BYTES):
        super().__init__(f"payload {actual} bytes exceeds limit {limit}")
        self.actual = actual
        self.limit = limit


class CanonicalMismatchError(RuntimeError):
    """Raised when a signing self-check fails.

    SPEC 03-modules-product §4 acceptance 3 requires the Python port
    to refuse to emit events whose canonical serialisation diverges
    from the TS reference. We cannot talk to Node at signing time, so
    we run a local self-check instead: after signing, the freshly
    produced event MUST verify under `verify_event` (which recomputes
    the canonical id from scratch). If it doesn't, the canonical
    bytes did not round-trip and we raise this typed error rather
    than emit an unverifiable event.
    """

    def __init__(self, reason: str):
        super().__init__(f"canonical self-check failed: {reason}")
        self.reason = reason


def _assert_kind(kind: str) -> None:
    if kind not in _PROTOCOL_KINDS:
        raise InvalidKindError(kind)


def create_event(
    kind: str,
    payload: dict[str, Any],
    tags: list[str],
    created_at: Optional[int] = None,
) -> dict[str, Any]:
    """Build an unsigned envelope. Mirrors createEvent() in TS."""
    _assert_kind(kind)
    return {
        "v": 1,
        "created_at": created_at if created_at is not None else int(time.time()),
        "kind": kind,
        "payload": payload,
        "tags": tags,
        "visibility": "public",
    }


def _payload_byte_length(payload: Any) -> int:
    # Matches TS: new TextEncoder().encode(JSON.stringify(payload)).length.
    # Python json.dumps with ensure_ascii=False produces the same UTF-8
    # byte count as JS JSON.stringify for the field types we emit.
    return len(json.dumps(payload, ensure_ascii=False).encode("utf-8"))


def sign_event(event: dict[str, Any], agent: AgentKey) -> dict[str, Any]:
    """Sign an unsigned envelope with an AgentKey."""
    _assert_kind(event["kind"])

    payload_bytes = _payload_byte_length(event["payload"])
    if payload_bytes > MAX_PAYLOAD_BYTES:
        raise PayloadTooLargeError(payload_bytes)

    with_keys = {
        **event,
        "pubkey": agent.public_key,
        "operator_pubkey": agent.delegated_by,
    }

    canonical = canonicalize(with_keys)
    event_id = sha256_hex(canonical)
    sig = sign_bytes(agent.private_key, bytes.fromhex(event_id)).hex()
    signed = {**with_keys, "id": event_id, "sig": sig}
    if not verify_event(signed):
        raise CanonicalMismatchError(
            "signed event failed local re-verification; refusing to emit"
        )
    return signed


def verify_event(event: dict[str, Any]) -> bool:
    """Recompute id and verify signature. Never raises."""
    try:
        rest = {k: v for k, v in event.items() if k not in ("id", "sig")}
        expected_id = sha256_hex(canonicalize(rest))
        if event["id"] != expected_id:
            return False
        return verify_bytes(
            event["pubkey"], bytes.fromhex(event["id"]), bytes.fromhex(event["sig"])
        )
    except Exception:
        return False
